sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobOutputStream.java

```diff
@@ -234,7 +234,8 @@ private BlockBlobOutputStream(final BlobAsyncClient client,
                     }
                     return Mono.empty();
                 })
-                .doOnTerminate(() -> {
+                // Use doFinally to cover all termination scenarios of the Flux.
+                .doFinally(signalType -> {
```

Comment:
Were we able to determine if `doOnTerminate` caused issues when a timeout occurred?

Reply:
I think so. When `timeout` operator is applied, the signal type is "cancel".

```java
// outer flux waiting for inner flux to complete.
Flux.defer(() -> {
    // inner flux timeout after 3 seconds
    return Flux.just(42)
            .delayElements(Duration.ofSeconds(10))
            .timeout(Duration.ofSeconds(3));
}).doFinally(signalType -> {
    System.out.println("doFinally:" + signalType);
})
.timeout(Duration.ofSeconds(3))
.blockLast();
```

output:

> doFinally:cancel

The `cancel` signal won't invoke `doOnTerminate`, which means in the sync BlobOutputStream::commit can lockup when cancelled.

Reply:
Now whether or not this is the problem customers are encountering is another question. I tried adding a test for BlobOS where I just added a pipeline policy that would timeout in a nanosecond I was still getting the completion signal at the uploadWithResponse layer